src/wesnoth.cpp

```diff
@@ -309,6 +309,10 @@ static int process_command_args(const commandline_options& cmdline_opts)
 		return 0;
 	}
 
+	// earliest possible point to ensure the userdata directory is known
+	// if you're hitting the assertion in the preferences about the userdata not being initialized, it means you're trying to use the preferences before wesnoth knows where the preferences are
+	filesystem::get_user_data_dir();
```

Comment:
I would rather do the actual check here and assert in `get_user_data_path`.

Reply:
I don't understand what you mean. What check and what assert?

Reply:
I assume you call `get_user_data_dir` since that ends up at
https://github.com/wesnoth/wesnoth/blob/e67a91caf22429c56cd2012e324cd8ee212b5481/src/filesystem.cpp#L908-L910
So rather explicitly check here and call `set_user_data_dir` and convert the check in `get_user_data_path` into an assert.

Then we have one single point where the user data dir is set up and `get_user_data_path` asserts if any code requests the userdata dir too early instead of doing a potentially bogus setup.

Reply:
This will cause the boost unit tests to fail, since they directly create instances of objects that depend on the preferences without going through normal startup.

Reply:
Why not fix that? Can you point to the code?

Reply:
To be clear this is an improvement no matter what. It's not crucial to get such an assert in in this PR.

Reply:
I think this is something I'd prefer be done separately since getting past the assert then just give another error and I don't want to have to work through whatever changes are needed for the boost tests in order to get the assert working as part of this PR.